npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,447 @@
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?)*$`
+    Name string `json:"names,omitempty"`
+
+    //  Pattern allows matching FQDNs using wildcard specifiers.
+    //
+    // "*" matches 0 or more DNS valid characters (except for "."), and may
+    // occur anywhere in the pattern.
+    //  Examples:
+    //    - `*.kubernetes.io` matches subdomains of kubernetes.io at that level.
+    //      "www.kubernetes.io" and "blog.kubernetes.io" match, however
+    //      "kubernetes.io", "latest.blog.kubernetes.io", and "wikipedia.org" do
+    //      not.
+    //    - `sub*.kubernetes.io` matches subdomains of kubernetes.io where the
+    //      subdomain component begins with "sub".
+    //      "sub.kubernetes.io" and "subdomain.kubernetes.io" match, however
+    //      "www.kubernetes.io", "blog.kubernetes.io", and "kubernetes.io do
+    //      not.
+    //    - `*.*.kubernetes.io` matches subdomains of kubernetes.io at that
+    //      level.
+    //      "sub.subdomain.kubernetes.io" matches, however "kubernetes.io" and
+    //      "blog.kubernetes.io" do not.
+    //    - `*a*.kubernetes.io` matches subdomains of kubernetes.io that contain
+    //      an "a" at that level.
+    //      "api.kubernetes.io" and "domain.kubernetes.io" match, however
+    //      "blog.kubernetes.io" and "kubernetes.io" do not
+    //
```

Comment:
I removed the more complicated regex matching examples. They aren't really necessary, I just got a little carried away with the matcher 😄 

I've left just *.kubernetes.io and *.*.kubernetes.io. I think that captures all the use-cases. I think the user-story for *.*.kubernetes.io is basically the same as for *.kubernetes.io, but for now `*` is just scoped to one subdomain level. Happy to revisit that.

Reply:
I guess we need to also change validation to not allow asterisks in random places. 
Also wildcard matching may be expensive, and allowing two-subdomain matches raises potential performance concerns, and also questions like: why only 2 subdomain and not 3? So maybe we could start with 1 wildcard match and then extend it in the future?

Reply:
it was also brought to my attention, that it may be better to only allow one wildcard label and match >0 labels instead of doing `*.*.domain.com` as per https://www.rfc-editor.org/rfc/rfc1034#section-4.3.3
```
Another way to look at this is that the "*" label always matches at least one whole
label and sometimes more, but always whole labels.
```

Reply:
I don't really have a strong preference one way or the other -- whether `*` matches 1 or multiple labels.

@Dyanngg @astoycos any thoughts?

Reply:
Hrm so I think one of these options
- Leave it up to the implementation to specify
OR
- Start with specifying that it matches only 1 label and expand in the future

(I'd probably go with the latter for now)

Reply:
Starting simple probably makes sense, but we should add a criteria for beta-graduation, saying that we need to revisit the semantics of `*`, since changing it is not backwards-compatible.

Reply:
Ok for now, it matches a single level -- we'll leave it that way. We can revisit upon graduation.